src/mmfatl.c

```diff
@@ -71,29 +153,110 @@ static void initBuffer(void) {
 
 bool test_initBuffer(void) {
   // emulate memory corruption
-  memset(buffer, 'x', sizeof(buffer));
+  memset(&buffer, 'x', sizeof(buffer));
 
   initBuffer();
 
+  ASSERT(buffer.begin == buffer.text)
   unsigned i = 0;
 
   // check the buffer is filled with NUL...
   for (;i < BUFFERSIZE; ++i)
-    ASSERT(buffer[i] == NUL);
+    ASSERT(buffer.text[i] == NUL);
+
+  ASSERT(buffer.end == buffer.text + i)
 
   // ...and has the ELLIPSIS string at the end...
   char ellipsis[] = ELLIPSIS;
   unsigned j = 0;
   for (;ellipsis[j] != NUL; ++i, ++j)
-    ASSERT(buffer[i] == ellipsis[j]);
+    ASSERT(buffer.text[i] == ellipsis[j]);
 
   // ... and a terminating NUL character
-  ASSERT(buffer[i] == NUL);
+  ASSERT(buffer.text[i] == NUL);
+  return true;
+}
+
+char const* bufferCompare(char const* match, int from, unsigned lg,
+                        unsigned begin)
+{
+  if (memcmp(buffer.begin + from, match, lg) != 0)
+    return "unexpected buffer contents";
+  return buffer.begin == buffer.text + begin?
+    (char const*)0 : "unexpected buffer begin";
```

Comment:
```suggestion
    (char const*)NULL : "unexpected buffer begin";
```
double check what the preferred way to write nullptr in C is